### PR TITLE
Implement custom cache path

### LINKED ISSUE #3425: Allow setting cache path
### Select Tool

snforge

### Requested Feature

Allow specifying caches path by env variable.

### PR DESCRIPTION
**Pull request by franciszekjob**
Closes #3425

## Introduced changes

Allow setting custom cache path via `SNFOUNDRY_CACHE` env.

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Comment by franciszekjob** [comment]
This change is part of the following stack:

- #4341 ◀
    - #4342

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

**Comment by franciszekjob** [comment]
note: I didn't mention the possibility of setting env in `clean-cache` docs, because the command removed in the next PR (due to deprecation).

File: crates/forge/src/clean.rs [hunk 1]
```diff
@@ -40,28 +40,25 @@ pub fn clean(args: CleanArgs, ui: &UI) -> Result<()> {
...
-        clean_dir(root_dir, dir_name, ui)?;
-    }
-    Ok(())
-}
-fn clean_dir(dir: &Utf8PathBuf, dir_name: &str, ui: &UI) -> Result<()> {
-    let dir = dir.join(dir_name);
-    if dir.exists() {
-        fs::remove_dir_all(&dir).with_context(|| format!("Failed to remove directory: {dir}"))?;
-        ui.println(&format!("Removed directory: {dir}"));
+fn clean_dir(path: &Utf8PathBuf, ui: &UI) -> Result<()> {
```
Comment (MKowalski8) [inline crates/forge/src/clean.rs]: nit: It's better to use `Utf8Path` if possible
```suggestion
fn clean_dir(path: &Utf8Path, ui: &UI) -> Result<()> {
```

File: crates/forge-runner/src/lib.rs [hunk 2]
```diff
@@ -42,7 +43,13 @@ pub mod test_case_summary;
 pub mod test_target_summary;
 pub mod tests_summary;
 
-pub const CACHE_DIR: &str = ".snfoundry_cache";
+pub const DEFAULT_CACHE_DIR: &str = ".snfoundry_cache";
+
+#[must_use]
+pub fn resolve_cache_dir(workspace_root: &Utf8Path) -> Utf8PathBuf {
+    let cache_dir = env::var("SNFOUNDRY_CACHE").unwrap_or_else(|_| DEFAULT_CACHE_DIR.to_string());
+    workspace_root.join(cache_dir)
```
Comment (MKowalski8) [inline crates/forge-runner/src/lib.rs]: Could we add a test that check absoulte path?

Or at lest add a comment here, sth like `// If SNFOUNDRY_CACHE is absolute, join() discards workspace_root`

For me it was not obvious that it work like this, so it might be helpful.

File: docs/src/appendix/snforge/clean.md [hunk 3]
```diff
@@ -8,7 +8,12 @@ Clean all generated directories
 
 ## `cache`
 
-Clean the `.snfoundry_cache` directory.
+Clean the cache directory.
+
+> ℹ️ **Info**
+>
+> The default cache directory is `.snfoundry_cache`.
+> You can override it with the `SNFOUNDRY_CACHE` environment variable.
```
Comment (MKowalski8) [inline docs/src/appendix/snforge/clean.md]: We could add also info that it is relative to workspace root. 

For example some env in cargo is relative to CWD, so it's not obvious. 

`CARGO_TARGET_DIR — Location of where to place all generated artifacts, relative to the current working directory. See [build.target-dir](https://doc.rust-lang.org/cargo/reference/config.html#buildtarget-dir) to set via config.`

https://doc.rust-lang.org/cargo/reference/environment-variables.html

BTW. We could add `Environment Variables` page in appendix as well, it will make things easier. Added issue for that #4347

File: crates/forge-runner/src/lib.rs (hunk 2 above)
Comment (ddoktorski) [inline crates/forge-runner/src/lib.rs]: I believe `SNFOUNDRY_CACHE` should always be an absolute path, otherwise we introduce a lot of ambiguity

File: crates/forge/tests/e2e/clean.rs [hunk 4]
```diff
@@ -156,6 +194,47 @@ fn test_clean_all() {
     );
 }
 
+#[test]
+#[cfg_attr(
+    feature = "cairo-native",
+    ignore = "Native doesn't support trace, coverage and profiler yet"
```
Comment (ddoktorski) [inline crates/forge/tests/e2e/clean.rs]: nit:
```suggestion
    ignore = "Native doesn't support trace, coverage and profiler"
```

File: crates/forge-runner/src/lib.rs (hunk 2 above)
Comment (franciszekjob) [reply crates/forge-runner/src/lib.rs]: I think comment should be enough. Added.

File: crates/forge/src/clean.rs [hunk 5]
```diff
@@ -40,28 +40,25 @@ pub fn clean(args: CleanArgs, ui: &UI) -> Result<()> {
...
-    Ok(())
-}
-fn clean_dir(dir: &Utf8PathBuf, dir_name: &str, ui: &UI) -> Result<()> {
-    let dir = dir.join(dir_name);
-    if dir.exists() {
-        fs::remove_dir_all(&dir).with_context(|| format!("Failed to remove directory: {dir}"))?;
-        ui.println(&format!("Removed directory: {dir}"));
+fn clean_dir(path: &Utf8Path, ui: &UI) -> Result<()> {
+    if path.exists() {
+        fs::remove_dir_all(path).with_context(|| format!("Failed to remove directory: {path}"))?;
```
Comment (ddoktorski) [inline crates/forge/src/clean.rs:70]: I think we should be more cautious when removing files in a custom path. What about first removing files that match the layout and then removing the dir if it's empty?

File: docs/src/appendix/snforge/clean.md (hunk 3 above)
Comment (franciszekjob) [reply docs/src/appendix/snforge/clean.md]: > BTW. We could add Environment Variables page in appendix as well, it will make things easier. Added issue for that https://github.com/foundry-rs/starknet-foundry/issues/4347

Good idea

File: crates/forge/tests/e2e/forking.rs [hunk 6]
```diff
@@ -102,11 +111,64 @@ fn with_clean_cache() {
...
+
+    let default_cache_dir = temp.path().join(DEFAULT_CACHE_DIR);
+    let custom_cache_dir = temp.path().join("custom_cache");
+    fs::create_dir(&custom_cache_dir).unwrap();
+
+    for entry in fs::read_dir(&default_cache_dir).unwrap() {
+        let entry = entry.unwrap();
+        fs::copy(entry.path(), custom_cache_dir.join(entry.file_name())).unwrap();
+    }
+    fs::remove_dir_all(&default_cache_dir).unwrap();
```
Comment (ddoktorski) [inline crates/forge/tests/e2e/forking.rs]:
```suggestion
    fs::rename(&default_cache_dir, &custom_cache_dir).unwrap();
```

**Review by MKowalski8 — APPROVED** [review]

**Review by ddoktorski — CHANGES_REQUESTED** [review]

File: crates/forge/src/clean.rs (hunk 5 above)
Comment (MKowalski8) [reply crates/forge/src/clean.rs:70]: I agree, I just noticed that too. Someone could unknowingly set their cache path to the `home directory` or so.

Comment (franciszekjob) [reply crates/forge/src/clean.rs:70]: Good catch. Changed and added a test which checks such scenario.

File: crates/forge/src/clean.rs [hunk 7]
```diff
@@ -31,6 +33,7 @@ pub fn clean(args: CleanArgs, ui: &UI) -> Result<()> {
         ..MetadataOpts::default()
     })?;
     let workspace_root = scarb_metadata.workspace.root;
+    let has_custom_cache_dir = env::var("SNFOUNDRY_CACHE").is_ok();
```
Comment (ddoktorski) [inline crates/forge/src/clean.rs]: It can be checked in `clean_cache_dir` function without passing it as an argument

File: crates/forge/src/clean.rs [hunk 8]
```diff
@@ -40,29 +43,96 @@ pub fn clean(args: CleanArgs, ui: &UI) -> Result<()> {
...
+        if is_snfoundry_cache_file(entry_path) {
+            fs::remove_file(entry_path)
+                .with_context(|| format!("Failed to remove cache file: {entry_path}"))?;
+            ui.println(&format!("Removed file: {entry_path}"));
+        }
+    }
+
+    let is_empty = path
+        .read_dir_utf8()
+        .with_context(|| format!("Failed to read cache directory: {path}"))?
```
Comment (ddoktorski) [inline crates/forge/src/clean.rs]: nit: We can avoid reading the same directory twice, one idea is to add a bool variable that will be set to true in the loop if there is any non snfoundry cache file:
```rust
  if is_snfoundry_cache_file(entry_path) {
      fs::remove_file(entry_path)
          .with_context(|| format!("Failed to remove cache file:
{entry_path}"))?;
      ui.println(&format!("Removed file: {entry_path}"));
  } else {
      any_skipped = true;
  }
```

File: crates/forge/src/clean.rs [hunk 9]
```diff
@@ -40,29 +43,96 @@ pub fn clean(args: CleanArgs, ui: &UI) -> Result<()> {
...
+
+    if is_empty {
+        fs::remove_dir(path).with_context(|| format!("Failed to remove directory: {path}"))?;
+        ui.println(&format!("Removed directory: {path}"));
     }
 
     Ok(())
 }
+
+fn is_snfoundry_cache_file(path: &Utf8Path) -> bool {
```
Comment (ddoktorski) [inline crates/forge/src/clean.rs:107]: Let's add some unit tests for this function

File: crates/forge-runner/src/lib.rs [hunk 10]
```diff
@@ -42,7 +43,21 @@ pub mod test_case_summary;
...
+    match env::var("SNFOUNDRY_CACHE") {
+        Ok(cache_dir) => {
+            let cache_dir = Utf8PathBuf::from(cache_dir);
+            ensure!(
+                cache_dir.is_absolute(),
+                "SNFOUNDRY_CACHE must be an absolute path"
+            );
+            Ok(cache_dir)
+        }
+        Err(_) => Ok(workspace_root.join(DEFAULT_CACHE_DIR)),
```
Comment (ddoktorski) [inline crates/forge-runner/src/lib.rs]: Shouldn't we distinguish between different errors `NotPresent` and `NotUnicode`? In the second case I think it should be a hard error

File: crates/forge/src/clean.rs [hunk 11]
```diff
@@ -40,29 +43,96 @@ pub fn clean(args: CleanArgs, ui: &UI) -> Result<()> {
...
+        return false;
+    };
+    let Some((sanitized_url, block_number)) = prefix.rsplit_once('_') else {
+        return false;
+    };
+
+    !sanitized_url.is_empty()
+        && !block_number.is_empty()
+        && block_number.chars().all(|char| char.is_ascii_digit())
+        && version.split('_').count() >= 3
```
Comment (ddoktorski) [inline crates/forge/src/clean.rs]: Why `>=` not `==`?

Comment (ddoktorski) [inline crates/forge/src/clean.rs]: Why not to use regex matching for this logic?

File: crates/forge/src/clean.rs [hunk 12]
```diff
@@ -1,13 +1,15 @@
...
-use camino::Utf8PathBuf;
+use camino::{Utf8Path, Utf8PathBuf};
+use forge_runner::resolve_cache_dir;
 use foundry_ui::UI;
 use scarb_api::metadata::{MetadataOpts, metadata_with_opts};
+use std::env;
 use std::fs;
 
 const COVERAGE_DIR: &str = "coverage";
+const FILE_WITH_PREV_TESTS_FAILED: &str = ".prev_tests_failed";
```
Comment (ddoktorski) [inline crates/forge/src/clean.rs]: This const is duplicated now within the same crate

File: crates/forge/src/clean.rs (hunk 11 above)
Comment (franciszekjob) [reply crates/forge/src/clean.rs]: c2ffdcb7cf8091e9b5ff05e07bd63f96fd1a235f

Comment (franciszekjob) [reply crates/forge/src/clean.rs]: Not relevant, cause we use regex now.

File: crates/forge-runner/src/lib.rs [hunk 13]
```diff
@@ -245,3 +264,48 @@ pub fn function_args(
...
+        assert_eq!(
+            resolve_cache_dir(workspace).unwrap(),
+            workspace.join(DEFAULT_CACHE_DIR)
+        );
+    }
+
+    #[test]
+    fn accepts_absolute_custom_path() {
+        // SAFETY: variable only modified within this test and cleaned up afterwards
+        unsafe { env::set_var("SNFOUNDRY_CACHE", "/var/cache/snfoundry") };
```
Comment (ddoktorski) [inline crates/forge-runner/src/lib.rs]: With this approach it seems that there can be race conditions as different tests require different cache dirs. What about splitting `resolve_cache_dir`  to public and private function, for example:
```rust
  pub fn resolve_cache_dir(workspace_root: &Utf8Path) -> Result<Utf8PathBuf> {
      resolve_cache_dir_impl(workspace_root, env::var("SNFOUNDRY_CACHE"))
  }
```
and then test here only `resolve_cache_dir_impl`?

File: crates/forge/src/clean.rs [hunk 14]
```diff
@@ -40,29 +51,122 @@ pub fn clean(args: CleanArgs, ui: &UI) -> Result<()> {
...
-    if dir.exists() {
-        fs::remove_dir_all(&dir).with_context(|| format!("Failed to remove directory: {dir}"))?;
-        ui.println(&format!("Removed directory: {dir}"));
+
+pub fn clean_cache_dir(path: &Utf8Path, ui: &UI) -> Result<()> {
+    if !path.exists() {
+        return Ok(());
+    }
+
+    let has_custom_cache_dir = env::var("SNFOUNDRY_CACHE").is_ok();
```
Comment (ddoktorski) [inline crates/forge/src/clean.rs]: To simplify this logic we can always check content as if it were a custom path. Alternatively `resolve_cache_dir` could return a tuple containing a boolean to indicate if it's a custom path

Comment (franciszekjob) [reply crates/forge/src/clean.rs]: Done, went with always checking content.
https://github.com/foundry-rs/starknet-foundry/pull/4341/commits/7257b5854d0bc2a0e749cb217e6034d51aa84684

File: crates/forge-runner/src/lib.rs (hunk 13 above)
Comment (franciszekjob) [reply crates/forge-runner/src/lib.rs]: https://github.com/foundry-rs/starknet-foundry/pull/4341/commits/26dd4fe69dc33afef99237183f22d33fcd709223

File: docs/src/appendix/snforge/clean.md [hunk 15]
```diff
@@ -8,7 +8,12 @@ Clean all generated directories
 
 ## `cache`
 
-Clean the `.snfoundry_cache` directory.
+Clean the cache directory.
+
+> ℹ️ **Info**
+>
+> The default cache directory is `.snfoundry_cache`, relative to the workspace root.
+> You can override it with the `SNFOUNDRY_CACHE` environment variable, which must be an absolute path.
```
Comment (ddoktorski) [inline docs/src/appendix/snforge/clean.md]: nit:
```suggestion
> You can override it with the `SNFOUNDRY_CACHE` environment variable, which must be an absolute path to a dedicated directory.
```

File: crates/forge-runner/src/lib.rs [hunk 16]
```diff
@@ -245,3 +271,39 @@ pub fn function_args(
         })
         .collect()
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (ddoktorski) [inline crates/forge-runner/src/lib.rs:276]: Let's add a test to make sure that it fails if `SNFOUNDRY_CACHE` is set to an empty string

File: crates/forge-runner/src/lib.rs [hunk 17]
```diff
@@ -245,3 +271,39 @@ pub fn function_args(
...
+    }
+
+    #[test]
+    fn rejects_relative_custom_path() {
+        let err = resolve_cache_dir_impl(
+            Utf8Path::new("/tmp/workspace"),
+            Ok("relative/cache".to_string()),
+        )
+        .unwrap_err();
+        assert!(err.to_string().contains("absolute path"));
```
Comment (ddoktorski) [inline crates/forge-runner/src/lib.rs]: nit: Assert exact full string

**Review by ddoktorski — APPROVED** [review]
